### PR TITLE
added restart method to action

### LINKED ISSUE #1249: Add Restart method to action
### Description of the task

In response to a bug found in PatrolTactic, where once a moveAction was done, we were expecting it to execute from the beginning with a different target destination. 

To prevent this, we should make a Restart method which causes an action to behave as if it were a new action (with the same control params). 
### Acceptance criteria

- [ ] Restart method implemented in `action.cpp` which sets `action->done()` to false
- [ ] Unit tests to cover this functionality

### PR DESCRIPTION
**Pull request by amrm3lm**
### Description
- added `restart()` method to action.cpp 
- Updated patrol_tactic to use `restart()` instead of recreating new action when it reaches a patrol point
 

### Testing Done

added unit test for restart() in action_test, also ran tests for patrol_tactic  
### Resolved Issues

resolves #1249 
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/action/action.h [hunk 1]
```diff
@@ -39,6 +39,11 @@ class Action
      */
     bool done() const;
 
+    /**
+     * runs the coroutine from the start, effectively restarting the action
```
Comment (garethellis0) [inline src/software/ai/hl/stp/action/action.h]: Hmm, I'd prefer not to mention "coroutine" in the interface because it's technically an implementation detail, but seems somewhat relevant here. Thoughts @MathewMacDougall @akhilveeraghanta @jonathanlew ?

Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/action.h]: Agree that I'd prefer not to mention it here if possible. I think we can just reword it to refer to the state, since Actions are known to be stateful.

eg. something like
~"Resets any state and restarts the Action from the beginning"~
This actually doesn't reset the state of the control variables so isn't entirely correct

How about just "Restarts the Action logic"

Do you think that sufficiently gets the message across? I originally had extra words like "from the beginning" but felt that was ultimately redundant.

Comment (garethellis0) [reply src/software/ai/hl/stp/action/action.h]: I think your suggestion here is good.

File: src/software/ai/hl/stp/action/action.h [hunk 2]
```diff
@@ -39,6 +39,11 @@ class Action
...
+     */
+    void restart();
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/action.h:45]:
```suggestion
    void restart(void);
```
better to be explicit

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good just a small nit.

File: src/software/ai/hl/stp/action/action.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/action.h:45]: counter nit: not saying `void` is also pretty explicit, and is consistent with 99% of the codebase. No preference here since it's the same and doesn't impact readability, but this should be in the style guide if it's something you want to consistently enforce.

Comment (jonathanlew) [reply src/software/ai/hl/stp/action/action.h:45]: https://stackoverflow.com/questions/51032/is-there-a-difference-between-foovoid-and-foo-in-c-or-c

There's a distinction in C  so that's why I'd prefer an explicit void but I agree it's not necessary in this case since it's C++

Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/action.h:45]: Ah TIL: https://softwareengineering.stackexchange.com/questions/286490/what-is-the-difference-between-function-and-functionvoid/286494

We should consider adding to our firmware / C style guide if it's not there already.

File: src/software/ai/hl/stp/action/action.h (hunk 1 above)
Comment (amrm3lm) [reply src/software/ai/hl/stp/action/action.h]: Agreed, changed comment to Matt's suggestion

File: src/software/ai/hl/stp/action/action.h (hunk 2 above)
Comment (amrm3lm) [reply src/software/ai/hl/stp/action/action.h:45]: Since theres no preference for c++ no changes were made from this

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Review by akhilveeraghanta — APPROVED** [review]
LGTM :tada::fire:

**Review by jonathanlew — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]
LGTM, great work! :+1:
